File: tools/oracle/smoke.c
#include <stdio.h>
#include "../../src/microui.h"

static int text_width(mu_Font font, const char *str, int len) {
  (void) font;
  if (len < 0) {
    len = 0;
    while (str[len]) { len++; }
  }
  return len;
}

static int text_height(mu_Font font) {
  (void) font;
  return 10;
}

int main(void) {
  mu_Context ctx;
  mu_Command *cmd = NULL;

  mu_init(&ctx);
  ctx.text_width = text_width;
  ctx.text_height = text_height;

  mu_begin(&ctx);
  if (mu_begin_window(&ctx, "Test", mu_rect(0, 0, 100, 100))) {
    int widths[] = { -1 };
    mu_layout_row(&ctx, 1, widths, 0);
    mu_label(&ctx, "Hello");
    mu_end_window(&ctx);
  }
  mu_end(&ctx);

  while (mu_next_command(&ctx, &cmd)) {
    switch (cmd->type) {
      case MU_COMMAND_CLIP:
        printf("clip %d %d %d %d\n", cmd->clip.rect.x, cmd->clip.rect.y, cmd->clip.rect.w, cmd->clip.rect.h);
        break;
      case MU_COMMAND_RECT:
        printf(
          "rect %d %d %d %d %u %u %u %u\n",
          cmd->rect.rect.x, cmd->rect.rect.y, cmd->rect.rect.w, cmd->rect.rect.h,
          cmd->rect.color.r, cmd->rect.color.g, cmd->rect.color.b, cmd->rect.color.a);
        break;
      case MU_COMMAND_TEXT:
        printf("text %d %d %s\n", cmd->text.pos.x, cmd->text.pos.y, cmd->text.str);
        break;
      case MU_COMMAND_ICON:
        printf("icon %d %d %d %d %d\n", cmd->icon.id, cmd->icon.rect.x, cmd->icon.rect.y, cmd->icon.rect.w, cmd->icon.rect.h);
        break;
    }
  }

  return 0;
}

File: tests/smoke.rs
use microui::{mu_begin, mu_begin_window, mu_end, mu_end_window, mu_label, mu_layout_row, mu_next_command, mu_rect, Command, Context};

fn text_width(_font: microui::Font, text: &str, len: i32) -> i32 {
    if len < 0 {
        text.len() as i32
    } else {
        len
    }
}

fn text_height(_font: microui::Font) -> i32 {
    10
}

#[test]
fn simple_window_command_stream_matches_fixture() {
    let mut ctx = Context::new();
    ctx.text_width = Some(text_width);
    ctx.text_height = Some(text_height);

    mu_begin(&mut ctx);
    if mu_begin_window(&mut ctx, "Test", mu_rect(0, 0, 100, 100)) != 0 {
        mu_layout_row(&mut ctx, 1, Some(&[-1]), 0);
        mu_label(&mut ctx, "Hello");
        mu_end_window(&mut ctx);
    }
    mu_end(&mut ctx);

    let mut cursor = None;
    let mut actual = String::new();
    while let Some(cmd) = mu_next_command(&ctx, &mut cursor) {
        match cmd {
            Command::Clip(c) => {
                actual.push_str(&format!("clip {} {} {} {}\n", c.rect.x, c.rect.y, c.rect.w, c.rect.h));
            }
            Command::Rect(c) => {
                actual.push_str(&format!(
                    "rect {} {} {} {} {} {} {} {}\n",
                    c.rect.x, c.rect.y, c.rect.w, c.rect.h, c.color.r, c.color.g, c.color.b, c.color.a
                ));
            }
            Command::Text(c) => {
                actual.push_str(&format!("text {} {} {}\n", c.pos.x, c.pos.y, c.text));
            }
            Command::Icon(c) => {
                actual.push_str(&format!("icon {} {} {} {} {}\n", c.id, c.rect.x, c.rect.y, c.rect.w, c.rect.h));
            }
            Command::Jump(_) => {}
        }
    }

    let expected = std::fs::read_to_string("tests/fixtures/smoke.txt").unwrap();
    assert_eq!(actual, expected);
}
